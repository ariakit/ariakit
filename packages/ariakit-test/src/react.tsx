import type { ReactElement } from "react";
import { StrictMode } from "react";
import * as ReactTestingLibrary from "@testing-library/react";
import { flushMicrotasks, nextFrame, wrapAsync } from "./__utils.js";

export * from "./index.js";

export interface RenderOptions
  extends Omit<ReactTestingLibrary.RenderOptions, "queries"> {
  strictMode?: boolean;
}

export async function render(ui: ReactElement, options?: RenderOptions) {
  const wrapper = (props: { children: ReactElement }) => {
    const Wrapper = options?.wrapper;
    const element = Wrapper ? <Wrapper {...props} /> : props.children;
    if (!options?.strictMode) return element;
    return <StrictMode>{element}</StrictMode>;
  };

  return wrapAsync(async () => {
    const { unmount } = ReactTestingLibrary.render(ui, { ...options, wrapper });
    await flushMicrotasks();
    await nextFrame();
    await flushMicrotasks();
    return unmount;
  });
}
