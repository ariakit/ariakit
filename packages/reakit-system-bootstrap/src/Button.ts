import { css, cx } from "emotion";
import { ButtonHTMLProps, ButtonOptions } from "reakit/Button/Button";
import { useBoxProps as usePaletteBoxProps } from "reakit-system-palette/Box";
import { useDarken } from "reakit-system-palette/utils/darken";
import { useContrast } from "reakit-system-palette/utils/contrast";
import { BootstrapBoxOptions } from "./Box";

export type BootstrapButtonOptions = BootstrapBoxOptions & ButtonOptions;

export function useButtonOptions({
  unstable_system: { fill = "opaque", palette = "primary", ...system } = {},
  ...options
}: BootstrapButtonOptions): BootstrapButtonOptions {
  return { unstable_system: { fill, palette, ...system }, ...options };
}

export function useButtonProps(
  { unstable_system }: BootstrapButtonOptions,
  htmlProps: ButtonHTMLProps = {}
): ButtonHTMLProps {
  const {
    style: { color, backgroundColor, borderColor = "transparent" },
  } = usePaletteBoxProps({ unstable_system });

  const hoverBackgroundColor = useDarken(
    backgroundColor || color,
    unstable_system.fill !== "opaque" ? 0 : 0.1
  );
  const activeBackgroundColor = useDarken(hoverBackgroundColor, 0.1);
  const hoverBorderColor = useDarken(backgroundColor || color, 0.2);
  const activeBorderColor = useDarken(hoverBorderColor, 0.1);
  const hoverColor = useContrast(hoverBackgroundColor);
  const activeColor = useContrast(activeBackgroundColor);

  const button = css`
    display: inline-flex;
    font-weight: 400;
    align-items: center;
    justify-content: center;
    user-select: none;
    padding: 0.375em 0.75em;
    line-height: 1.5;
    border-radius: 0.25rem;
    text-decoration: none;
    border: 1px solid ${borderColor};
    cursor: pointer;
    white-space: nowrap;
    color: ${color};
    background-color: ${backgroundColor || "transparent"};
    transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
      border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
    font-size: 100%;

    &[aria-disabled="true"] {
      cursor: auto;
    }

    &:not([aria-disabled="true"]) {
      &:hover {
        color: ${hoverColor};
        border-color: ${hoverBorderColor};
        background-color: ${hoverBackgroundColor};
      }
      &:active,
      &[data-active],
      &[aria-expanded="true"] {
        color: ${activeColor};
        border-color: ${activeBorderColor};
        background-color: ${activeBackgroundColor};
      }
    }
  `;

  return { ...htmlProps, className: cx(button, htmlProps.className) };
}
