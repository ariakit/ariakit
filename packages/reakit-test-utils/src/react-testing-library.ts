export {
  act,
  fireEvent,
  render,
  wait,
  waitForDomChange,
  waitForElement,
  waitForElementToBeRemoved
} from "@testing-library/react";
