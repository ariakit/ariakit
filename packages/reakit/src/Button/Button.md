`Button` is built from [Box](/components/primitives/box) with minimal default styles to support the many possible uses a button has.

It renders a `<div>` by default since that's more flexible than `<button>`. Proper ARIA attributes and event handlers are applied in order to guarantee accessibility (see the `HTML` tab below).

```jsx
<Button>Button</Button>
```

It has `display: inline-grid` and `grid-gap: 0.68em` by default, which means that you can put icons and other elements inside the button and they will be properly distributed within it. Here's an example using [react-icons](https://github.com/react-icons/react-icons):

```jsx
import Bell from "react-icons/lib/fa/bell";

<Button>
  <Bell />Bell
</Button>
```
