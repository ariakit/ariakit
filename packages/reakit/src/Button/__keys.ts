// Automatically generated
export const BUTTON_KEYS = [] as const;
