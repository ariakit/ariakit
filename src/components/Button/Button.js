import styled from "styled-components";
import { prop } from "styled-tools";
import as from "../../enhancers/as";
import Box from "../Box";

const handleKeyPress = e => {
  if (e.charCode === 32 || e.charCode === 13) {
    e.preventDefault();
    e.target.click();
  }
};

const Button = styled(Box)`
  display: inline-flex;
  position: relative;
  flex: none;
  appearance: none;
  user-select: none;
  outline: none;
  align-items: center;
  white-space: nowrap;
  justify-content: center;
  text-decoration: none;
  color: inherit;
  cursor: pointer;
  background-color: rgba(0, 0, 0, 0.03);
  height: 2.5em;
  min-width: 2.5em;
  padding: 0 0.68em;
  &:hover,
  &:focus {
    box-shadow: inset 0 0 999em rgba(0, 0, 0, 0.1);
  }
  &:active,
  &.active {
    box-shadow: inset 0 0 999em rgba(0, 0, 0, 0.2);
  }
  &:after {
    display: none;
    content: "";
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    border-radius: inherit;
    background-color: rgba(255, 255, 255, 0.35);
  }
  &[disabled] {
    pointer-events: none;
    &:after {
      display: block;
    }
  }
  &:not(button):not(select):not(input) {
    display: inline-grid;
    grid-gap: 0.68em;
    grid-auto-flow: column;
    align-content: center;
  }
  ${prop("theme.Button")};
`;

Button.defaultProps = {
  role: "button",
  tabIndex: 0,
  onKeyPress: handleKeyPress
};

export default as("div")(Button);
