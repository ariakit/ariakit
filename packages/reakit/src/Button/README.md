---
path: /docs/button/
redirect_from:
  - /components/button/
---

# Button

Accessible `Button` component that enables users to trigger an action or event, such as submitting a [Form](/docs/form/), opening a [Dialog](/docs/dialog/), canceling an action, or performing a delete operation. It follows the [WAI-ARIA Button Pattern](https://www.w3.org/TR/wai-aria-practices/#button).

<carbon-ad></carbon-ad>

## Installation

```sh
npm install reakit
```

Learn more in [Get started](/docs/get-started/).

## Usage

```jsx
import { Button } from "reakit/Button";

function Example() {
  return <Button>Button</Button>;
}
```

## Styling

Reakit components are unstyled by default. You're free to use whatever approach you want. Each component returns a single HTML element that accepts all HTML props, including `className` and `style`.

> The example below uses [Emotion](https://emotion.sh/docs/introduction). But these styles can be reproduced using static CSS and other CSS-in-JS libraries, such as [styled-components](https://styled-components.com/).

```jsx unstyled
import { Button } from "reakit/Button";
import { css } from "emotion";

const className = css`
  outline: 0;
  color: #ffffff;
  background: #006dff;
  padding: 0.375em 0.75em;
  line-height: 1.5;
  border: transparent;
  border-radius: 0.25rem;
  cursor: pointer;
  font-size: 16px;

  &:focus {
    box-shadow: 0 0 0 0.2em rgba(0, 109, 255, 0.4);
  }

  &[disabled],
  &[aria-disabled="true"] {
    cursor: auto;
    opacity: 0.5;
  }

  &:not([disabled]),
  &:not([aria-disabled="true"]) {
    &:hover {
      color: #ffffff;
      background-color: #0062e6;
    }
    &:active,
    &[data-active="true"] {
      color: #ffffff;
      background-color: #004eb8;
    }
  }
`;

function Example() {
  return <Button className={className}>Button</Button>;
}
```

Learn more in [Styling](/docs/styling/).

## Accessibility

- `Button` has role `button`.
- When `Button` has focus, <kbd>Space</kbd> and <kbd>Enter</kbd> activates it.

  <!-- eslint-disable no-alert -->

  ```jsx
  import { Button } from "reakit/Button";

  function Example() {
    return (
      <Button as="div" onClick={() => alert("clicked")}>
        Button
      </Button>
    );
  }
  ```

- If `disabled` prop is `true`, `Button` has `disabled` and `aria-disabled` attributes set to `true`.

  <!-- eslint-disable no-alert -->

  ```jsx
  import { Button } from "reakit/Button";

  function Example() {
    return (
      <Button disabled onClick={() => alert("clicked")}>
        Button
      </Button>
    );
  }
  ```

- If `disabled` and `focusable` props are `true`, `Button` has `aria-disabled` attribute set to `true`, but not `disabled`.

  <!-- eslint-disable no-alert -->

  ```jsx
  import { Button } from "reakit/Button";

  function Example() {
    return (
      <Button disabled focusable onClick={() => alert("clicked")}>
        Button
      </Button>
    );
  }
  ```

  This is useful when the presence of a `Button` is important enough so users can perceive it by tabbing.

Learn more in [Accessibility](/docs/accessibility/).

## Composition

- `Button` uses [Clickable](/docs/clickable/), and is used by [FormPushButton](/docs/form/), [FormRemoveButton](/docs/form/), [Disclosure](/docs/disclosure/) and all their derivatives.

Learn more in [Composition](/docs/composition/#props-hooks).

## Props

<!-- Automatically generated -->

### `Button`

- **`disabled`**
  <code>boolean | undefined</code>

  Same as the HTML attribute.

- **`focusable`**
  <code>boolean | undefined</code>

  When an element is `disabled`, it may still be `focusable`. It works
similarly to `readOnly` on form elements. In this case, only
`aria-disabled` will be set.
