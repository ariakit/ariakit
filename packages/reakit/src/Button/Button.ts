import * as React from "react";
import { createComponent } from "reakit-system/createComponent";
import { createHook } from "reakit-system/createHook";
import { useForkRef } from "reakit-utils/useForkRef";
import { isButton } from "reakit-utils/isButton";
import { warning } from "reakit-warning";
import {
  ClickableOptions,
  ClickableHTMLProps,
  useClickable,
} from "../Clickable/Clickable";
import { BUTTON_KEYS } from "./__keys";

export type ButtonOptions = ClickableOptions;

export type ButtonHTMLProps = ClickableHTMLProps &
  React.ButtonHTMLAttributes<any>;

export type ButtonProps = ButtonOptions & ButtonHTMLProps;

export const useButton = createHook<ButtonOptions, ButtonHTMLProps>({
  name: "Button",
  compose: useClickable,
  keys: BUTTON_KEYS,

  useProps(_, { ref: htmlRef, ...htmlProps }) {
    const ref = React.useRef<HTMLElement>(null);
    const [role, setRole] = React.useState<"button" | undefined>(undefined);
    const [type, setType] = React.useState<"button" | undefined>("button");

    React.useEffect(() => {
      const element = ref.current;
      if (!element) {
        warning(
          true,
          "Can't determine whether the element is a native button because `ref` wasn't passed to the component",
          "See https://reakit.io/docs/button"
        );
        return;
      }
      if (!isButton(element)) {
        if (element.tagName !== "A") {
          setRole("button");
        }
        setType(undefined);
      }
    }, []);

    return {
      ref: useForkRef(ref, htmlRef),
      role,
      type,
      ...htmlProps,
    };
  },
});

export const Button = createComponent({
  as: "button",
  memo: true,
  useHook: useButton,
});
