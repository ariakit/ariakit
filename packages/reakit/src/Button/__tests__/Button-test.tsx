import * as React from "react";
import { render } from "react-testing-library";
import Button from "../Button";

test("html attrs", () => {
  const { getByTestId } = render(
    <Button id="test" aria-label="test" data-testid="test" />
  );

  expect(getByTestId("test")).toHaveAttribute("id", "test");
  expect(getByTestId("test")).toHaveAttribute("aria-label", "test");
});

test("styled", () => {
  const { container } = render(<Button />);
  expect(container.firstChild).toMatchSnapshot();
});
