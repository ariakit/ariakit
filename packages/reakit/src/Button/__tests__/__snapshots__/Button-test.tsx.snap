// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled 1`] = `
.c0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

<button
  class="c0"
/>
`;
